pulsar-client/src/main/java/org/apache/pulsar/client/impl/TypedMessageBuilderImpl.java

```diff
@@ -105,14 +106,12 @@ public CompletableFuture<MessageId> sendAsync() {
 
     @Override
     public TypedMessageBuilder<T> key(String key) {
-        if (schema.getSchemaInfo().getType() == SchemaType.KEY_VALUE) {
-            KeyValueSchemaImpl kvSchema = (KeyValueSchemaImpl) schema;
-            checkArgument(kvSchema.getKeyValueEncodingType() != KeyValueEncodingType.SEPARATED,
-                    "This method is not allowed to set keys when in encoding type is SEPARATED");
-            if (key == null) {
-                msgMetadata.setNullPartitionKey(true);
-                return this;
-            }
+        getKeyValueSchema().ifPresent(keyValueSchema -> checkArgument(
```

Comment:
While you're here, can you explain why when the schema is KeyValue, and the encoding is SEPARATED, it is forbidden to use this method to se the key of the message?

Reply:
Because when the encoding is SEPARATED, the partition key should be set by a `KeyValue` object.

```java
Producer<KeyValue<String, String>> producer = client.newProducer(
                Schema.KeyValue(Schema.STRING, Schema.STRING, KeyValueEncodingType.SEPARATED))
        .topic("my-topic")
        .create();
producer.newMessage().value(new KeyValue<>("key", "value")).send()
```

With the SEPARATED key value encoding type, the code above will send a message whose partition key is "key" and value is "value". If it's allowed to set the partition key here, conflicts will happen.

There is no design of the key value schema, I just guessed the reason why `key` and `keyBytes` methods are forbidden and followed the previous behavior in this PR